From 19d1c1d9e998979d8402e4c2cf7910149bb6bf5c Mon Sep 17 00:00:00 2001
From: AsavarTzeth <asavartzeth@gmail.com>
Date: Wed, 2 Sep 2020 14:54:18 +0200
Subject: [PATCH 2/4] Support reverse-dns filenames for data files

Flatpak isolates application files from the host, but some files must be
exported to the host to serve their purpose (ex. desktop entries, icons
mime types, etc...). Filenames must however be prefixed by the app id, in
reverse-dns notation (org.keepassxc.KeePassXC).

If KEEPASSXC_DIST_TYPE=Flatpak is set CMake can now install files with
exportable filenames, without affecting other distribution types.
---
 share/CMakeLists.txt                          | 47 +++++++++++++++----
 share/linux/.gitignore                        |  3 ++
 .../linux/{keepassxc.xml => keepassxc.xml.in} |  2 +-
 ...top => org.keepassxc.KeePassXC.desktop.in} |  2 +-
 src/gui/Icons.cpp                             | 10 ++++
 5 files changed, 53 insertions(+), 11 deletions(-)
 create mode 100644 share/linux/.gitignore
 rename share/linux/{keepassxc.xml => keepassxc.xml.in} (85%)
 rename share/linux/{org.keepassxc.KeePassXC.desktop => org.keepassxc.KeePassXC.desktop.in} (99%)

diff --git a/share/CMakeLists.txt b/share/CMakeLists.txt
index 6d689df9..b069e4a4 100644
--- a/share/CMakeLists.txt
+++ b/share/CMakeLists.txt
@@ -23,15 +23,44 @@ install(FILES ${wordlists_files} DESTINATION ${DATA_INSTALL_DIR}/wordlists)
 file(COPY "wordlists" DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
 
 if(UNIX AND NOT APPLE AND NOT HAIKU)
-    install(DIRECTORY icons/application/ DESTINATION ${CMAKE_INSTALL_DATADIR}/icons/hicolor
-            FILES_MATCHING PATTERN "keepassx*.png" PATTERN "keepassx*.svg"
-            PATTERN "status" EXCLUDE PATTERN "actions" EXCLUDE PATTERN "categories" EXCLUDE)
-    install(DIRECTORY icons/application/ DESTINATION ${CMAKE_INSTALL_DATADIR}/icons/hicolor
-            FILES_MATCHING PATTERN "application-x-keepassxc.png" PATTERN "application-x-keepassxc.svg"
-            PATTERN "status" EXCLUDE PATTERN "actions" EXCLUDE PATTERN "categories" EXCLUDE)
-    install(FILES linux/org.keepassxc.KeePassXC.desktop DESTINATION ${CMAKE_INSTALL_DATADIR}/applications)
-    install(FILES linux/org.keepassxc.KeePassXC.appdata.xml DESTINATION ${CMAKE_INSTALL_DATADIR}/metainfo)
-    install(FILES linux/keepassxc.xml DESTINATION ${CMAKE_INSTALL_DATADIR}/mime/packages)
+    # Flatpak requires all host accessible files to use filenames based upon the app id
+    if(KEEPASSXC_DIST_FLATPAK)
+        set(APP_ICON "${ID}")
+        set(MIME_ICON "${ID}-application-x-keepassxc")
+        configure_file(linux/keepassxc.xml.in linux/${ID}.xml @ONLY)
+        install(FILES linux/${ID}.xml DESTINATION ${CMAKE_INSTALL_DATADIR}/mime/packages)
+
+        file(GLOB_RECURSE ICON_FILES LIST_DIRECTORIES false
+             "icons/application/*/keepassxc*.png"
+             "icons/application/*/*keepassxc*.svg")
+        foreach(icon_match ${ICON_FILES})
+            get_filename_component(icon_name ${icon_match} NAME)
+            get_filename_component(icon_dir ${icon_match} DIRECTORY)
+            # Prefix all icons with application id: "org.keepassxc.KeePassXC"
+            string(REGEX REPLACE "^keepassxc(.*)?(\\.png|\\.svg)$" "${ID}\\1\\2" icon_name ${icon_name})
+            string(REGEX REPLACE "^(application-x-keepassxc\\.svg)$" "${ID}-\\1" icon_name ${icon_name})
+	    #string(REGEX REPLACE "^(${ID})-(.*)$" "\\1.\\2" icon_name ${icon_name})
+            # Find icon sub dir ex. "scalable/mimetypes/"
+            file(RELATIVE_PATH icon_subdir ${CMAKE_CURRENT_SOURCE_DIR}/icons/application ${icon_dir})
+            install(FILES ${icon_match} DESTINATION ${CMAKE_INSTALL_DATADIR}/icons/hicolor/${icon_subdir}
+                    RENAME ${icon_name})
+        endforeach()
+    else()
+        set(APP_ICON "keepassxc")
+        set(MIME_ICON "application-x-keepassxc")
+        configure_file(linux/keepassxc.xml.in keepassxc.xml @ONLY)
+        install(FILES linux/keepassxc.xml DESTINATION ${CMAKE_INSTALL_DATADIR}/mime/packages)
+
+        install(DIRECTORY icons/application/ DESTINATION ${CMAKE_INSTALL_DATADIR}/icons/hicolor
+                FILES_MATCHING PATTERN "keepassx*.png" PATTERN "keepassx*.svg"
+                PATTERN "status" EXCLUDE PATTERN "actions" EXCLUDE PATTERN "categories" EXCLUDE)
+        install(DIRECTORY icons/application/ DESTINATION ${CMAKE_INSTALL_DATADIR}/icons/hicolor
+                FILES_MATCHING PATTERN "application-x-keepassxc.svg" PATTERN "status"
+                EXCLUDE PATTERN "actions" EXCLUDE PATTERN "categories" EXCLUDE)
+    endif(KEEPASSXC_DIST_FLATPAK)
+    configure_file(linux/${ID}.desktop.in linux/${ID}.desktop @ONLY)
+    install(FILES linux/${ID}.desktop DESTINATION ${CMAKE_INSTALL_DATADIR}/applications)
+    install(FILES linux/${ID}.appdata.xml DESTINATION ${CMAKE_INSTALL_DATADIR}/metainfo)
 endif(UNIX AND NOT APPLE AND NOT HAIKU)
 
 if(APPLE)
diff --git a/share/linux/.gitignore b/share/linux/.gitignore
new file mode 100644
index 00000000..83528a14
--- /dev/null
+++ b/share/linux/.gitignore
@@ -0,0 +1,3 @@
+keepassxc.xml
+org.keepassxc.KeePassXC.xml
+org.keepassxc.KeePassXC.desktop
diff --git a/share/linux/keepassxc.xml b/share/linux/keepassxc.xml.in
similarity index 85%
rename from share/linux/keepassxc.xml
rename to share/linux/keepassxc.xml.in
index b26b4db2..15ec1118 100644
--- a/share/linux/keepassxc.xml
+++ b/share/linux/keepassxc.xml.in
@@ -3,6 +3,6 @@
   <mime-type type="application/x-keepass2">
     <comment>KeePass 2 Database</comment>
     <glob pattern="*.kdbx"/>
-    <icon name="application-x-keepassxc"/>
+    <icon name="@MIME_ICON@"/>
   </mime-type>
 </mime-info>
diff --git a/share/linux/org.keepassxc.KeePassXC.desktop b/share/linux/org.keepassxc.KeePassXC.desktop.in
similarity index 99%
rename from share/linux/org.keepassxc.KeePassXC.desktop
rename to share/linux/org.keepassxc.KeePassXC.desktop.in
index e8d4dccf..5b7779b4 100644
--- a/share/linux/org.keepassxc.KeePassXC.desktop
+++ b/share/linux/org.keepassxc.KeePassXC.desktop.in
@@ -37,7 +37,7 @@ Comment[et]=Kogukonna arendatav port Windowsi programmist KeePass Password Safe
 Comment[ru]=Разработанный сообществом порт Windows-приложения KeePass Password Safe
 Exec=keepassxc %f
 TryExec=keepassxc
-Icon=keepassxc
+Icon=@APP_ICON@
 StartupWMClass=keepassxc
 StartupNotify=true
 Terminal=false
diff --git a/src/gui/Icons.cpp b/src/gui/Icons.cpp
index 0809e271..73da3487 100644
--- a/src/gui/Icons.cpp
+++ b/src/gui/Icons.cpp
@@ -54,7 +54,11 @@ Icons::Icons()
 
 QIcon Icons::applicationIcon()
 {
+#ifdef KEEPASSXC_DIST_FLATPAK
+    return icon("org.keepassxc.KeePassXC", false);
+#else
     return icon("keepassxc", false);
+#endif
 }
 
 QString Icons::trayIconAppearance() const
@@ -81,7 +85,11 @@ QIcon Icons::trayIcon(QString style)
 
     auto iconApperance = trayIconAppearance();
     if (!iconApperance.startsWith("monochrome")) {
+#ifdef KEEPASSXC_DIST_FLATPAK
+	return icon(QString("org.keepassxc.KeePassXC%1").arg(style), false);
+#else
         return icon(QString("keepassxc%1").arg(style), false);
+#endif
     }
 
     QIcon i;
@@ -91,6 +99,8 @@ QIcon Icons::trayIcon(QString style)
     } else {
         i = icon(QString("keepassxc-monochrome-dark%1").arg(style), false);
     }
+#elif defined(KEEPASSXC_DIST_FLATPAK)
+    i = icon(QString("org.keepassxc.KeePassXC-%1%2").arg(iconApperance, style), false);
 #else
     i = icon(QString("keepassxc-%1%2").arg(iconApperance, style), false);
 #endif
-- 
2.35.1

